### PR TITLE
Allow to create local RRef with value

### PR DESCRIPTION
**Pull request by aazzolini**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28948 Allow to create local RRef with value**

Add the constructor RRef(value) in python. This allows to wrap a local object with RRef an pass or return this RRef to users.
This enables returning, for example, a list of RRefs containing the parameters of a module to the user of the module.

Differential Revision: [D18241227](https://our.internmc.facebook.com/intern/diff/D18241227/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rref_context.cpp]: It seems we need a lock here to guard `owners_`.

File: torch/distributed/rpc/api.py
Comment (mrshenli) [inline torch/distributed/rpc/api.py]: Why do we need `as PyRRef` here?

File: torch/csrc/distributed/rpc/init.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/init.cpp]: Move this to top as this is a constructor?

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rref_context.cpp]: I assume that this does not take a `T` argument as we can also use this to support run `rpc.remote` on self?

File: torch/distributed/rpc/api.py
Comment (aazzolini) [reply torch/distributed/rpc/api.py]: I wanted to expose RRef to rpc.RRef  -- but avoid the lint issues for RRef not being used in the file. I can alternatively add a lint ignore rule. Is this how you'd do it?

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (aazzolini) [reply torch/csrc/distributed/rpc/rref_context.cpp]: I agree that we can use this to run rpc.remote on self, but I don't follow how it would be possible to remove the T argument -- I need to pass it to the constructor of OnwerRRef.

File: torch/distributed/rpc/api.py
Comment (mrshenli) [reply torch/distributed/rpc/api.py]: Yep, either adding a `# noqa: F401` or do something similar to this PR #25823

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/rref_context.cpp]: Oh, I mean I was thinking this API should take a `T&& value` for now as this is how we use it currently, i.e., something like:

```c++
template <typename T>
std::shared_ptr<OwnerRRef<T>> RRefContext::createOwnerRRef(T&& value) {
```

Comment (aazzolini) [reply torch/csrc/distributed/rpc/rref_context.cpp]: Yes i'm thinking of a use case where we don't set the value right away.

File: test/rpc_test.py
Comment (pritamdamania87) [inline test/rpc_test.py]: This API is a little weird. This RRef has nothing to do with rpc, yet we're creating it with `rpc.RRef()`. I think RRef should be a separate package since it is independent from RPC (as we can see here). If we want to allow creating local RRefs like this, RRef shouldn't be tightly coupled with RPC.

File: torch/csrc/distributed/rpc/init.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/init.cpp]: Just realized that these are all part of the public user API and we should have good documentation for all of these. We should define all of these methods in api.py and have proper python docs. Don't have to do this as part of this PR.

File: torch/csrc/distributed/rpc/py_rref.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/py_rref.cpp]: We do we have a static method like this instead of having a constructor?

File: test/rpc_test.py
Comment (aazzolini) [reply test/rpc_test.py]: There's 2 things in your comment:

1. would you prefer to have simply torch.distributed.RRef ?  I can do that.

2. Currently init_rpc() needs to be called before we can call RRef() constructor even locally since only OnwerRRef constructor is private and only accessible by RRefContext. I can change that too if you think is appropriate? In that case, RRefs would become known to RRefContext only at fork time.

File: torch/csrc/distributed/rpc/py_rref.cpp
Comment (aazzolini) [reply torch/csrc/distributed/rpc/py_rref.cpp]: The constructor of PyRRef is not exposed to python. This is something that works on top of that constructor.

Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/py_rref.cpp]: Can't we define a constructor for PyRRef that takes a py::object and use that in py::init()?

File: test/rpc_test.py
Comment (pritamdamania87) [reply test/rpc_test.py]: 1. Yes, I think that makes sense. 
2. I think this is why we probably need a higher level initialization like `init_model_parallel` (I'm open to a better name) that initializes rpc, rref and autograd. The user has to call this API even if they want to only use rpc/rref/dist_autograd individually.

Comment (aazzolini) [reply test/rpc_test.py]: I thought a bit about this -- I don't think that it makes sense to initialize RRef without initializing RPC, actually. RRef semantic requires RPC concepts such as owner() . owner() only makes sense if an RPCAgent is present.

RPC without RRef doesn't make sense either -- since RRef is a native type supported as both input and output of any RPC calls, particularly Remote.

As for 1., I'll change it.

Comment (pietern) [reply test/rpc_test.py]: Please don't do 1. The top level `torch.distributed` package concerns only collectives. As you say, RPC without RRef and RRef without RPC are both useless. If that ever changes, we could consider promoting RRefs to their own package, but not to the top level `torch.distributed` module.

Comment (pietern) [reply test/rpc_test.py]: Update: I see they're part of `torch.distributed` today, but that's because that's where they were defined from the very beginning. I have #27529 out (pending type checking fixes) to update that and move the pybind11 wrappers to the appropriate modules. We can discuss moving them around, but let's use a new issue for that.

Comment (aazzolini) [reply test/rpc_test.py]: Ok i'm just going to reuse the existing distributed.RRef for now. Will leave to @pietern to move it afterwards.

**Comment by pietern** [comment]
@aazzolini Please hold off on merging until #27529 has landed. I've been trying to merge it for a while and it should be any moment now...

**Comment by pietern** [comment]
@aazzolini The PR has been merged. If you rebase you can use `rpc.RRef`. See #27529 and b4df413712098ff93da694472d95896e62faf07d.

File: test/rpc_test.py
Comment (mrshenli) [inline test/rpc_test.py]: This is currently flaky. Can you explicitly hold the rref alive for now?

**Review by mrshenli — APPROVED** [review]
Looks good from my side! Thanks!

File: test/rpc_test.py
Comment (mrshenli) [reply test/rpc_test.py]: I am landing #29396, which will revert `to_here()` and make it block again.

**Pull request closed without merging**
